src/posts/render-blocking-resources.md

```diff
@@ -0,0 +1,247 @@
+
+Understanding render-blocking resources will enable you to fix this common web performance issue. In this post, you will learn:
+
+- What render-blocking resources are
+- Why they are important to performance
+- How to test and measure your website
+- How to fix this issue
+
+Before we continue, we need to take a step back and understand what the critical rendering path is.
+
+## What is the critical rendering path?
+
+We write HTML, CSS, and JavaScript in files and then deliver those files to the browser. The browser converts those files into the page you see through the **critical rendering path**. The steps are:
+
+1. Download the **HTML**
+2. Read the HTML, and concurrently:
+    - Construct the **Document Object Model** (DOM)
+    - Notice a `<link>` tag for a stylesheet and download the **CSS**
+3. Read the CSS and construct the **CSS Object Model** (CSSOM)
+4. Combine the DOM and the CSSOM into a **render tree**
+5. Using the render tree, compute the **layout** (size and position of each element)
+6. **Paint**, or render, the pixels on the page
+
+<figure id="critical-render-path">
+  <img src="/img/critical_render_path_sia_karamalegos.svg"
+    alt="Steps of the critical render path visualized in a diagram"
+    width="595" height="707">
+  <figcaption>The critical render path (<a href="#critical-render-path">link</a>)</figcaption>
+</figure>
+
+We want this process to go as fast as possible. Can you guess what makes it go slower?
+
+{% include 'newsletter-aside.njk' %}
+
+## What are render-blocking resources?
+
+Render-blocking resources are files that "press pause" on the critical rendering path. They interrupt one or more of the steps.
+
+HTML is technically render blocking because you need it to create the DOM. Without the HTML we would not even have a page to render.
+
+However, HTML is not usually the cause of our problems...
+
+**CSS is render blocking**. The browser needs it before it can create the CSSOM, which blocks all later steps. As soon as the browser encounters a stylesheet `<link>` or `<style>` tag, it must download and parse the contents. Then it must create the CSSOM before the rest of the render can continue. You can see this represented at the triangle point in the [diagram](#critical-render-path). The render tree cannot continue until both the CSSOM and DOM are created.
+
+**JavaScript CAN be render blocking**. It blocks the critical rendering path when it interacts with either the DOM or CSSOM. When the browser encounters a script meant to run synchronously, it will stop DOM creation until the script is finished running:
+
+<figure id="critical-render-path-js">
+  <img src="/img/critical_render_path_JS_karamalegos.svg"
+    alt="HTML encounters a synchronous script in the head which stops the parser"
+    width="595" height="862">
+  <figcaption>Synchronous JavaScript (no async or defer) will block the HTML parser both for download and execution (<a href="#critical-render-path-js">link</a>)</figcaption>
```

Comment:
I love the diagrams but it's inaccurate because the CSS will be downloaded before the script executes. This is because of speculative parsing that will look ahead to discover resources, even if the HTML parser is blocked.

It's super tricky to represent that in a diagram though.

https://andydavies.me/blog/2013/10/22/how-the-browser-pre-loader-makes-pages-load-faster/
https://webkit.org/blog/166/optimizing-page-loading-in-web-browser/
https://github.com/whatwg/html/issues/5624

Reply:
Thanks! I have an idea for that. I'm wondering if that means the CSSOM construction is not blocked? I feel like it should be though.

Reply:
I have not explored this deeply enough recently to know the answer, sorry.

Reply:
Thanks again. Just edited the diagram and pushed up all the changes.

Reply:
That's a good diagram: more accurate without going too deep into speculative parsing.

I think the `<figcaption>` still needs updating though.

Reply:
Great catch, thanks so much!

Reply:
Seems like you've updated the other figcaption (and that's a good edit). This one still mentions that it blocks downloads.

Reply:
I presumed a non-async script would block the speculative parser as well as the HTML parser because the script might change the content?

Reply:
Nope, the speculative parser was created exactly for sync scripts and in particular `document.write()` (see the links I shared). It is called speculative because it could download resources that won't be needed once the real parser as done its job. Bugs aside, the idea works well in practice because we rarely use `document.write` to comment out a section of HTML.

Reply:
For this one I actually meant the parser is blocked during download and execution of the JavaScript. Not that it blocks other files for download. Is this more clear?

> Synchronous JavaScript (no async or defer) will block the HTML parser during both download and execution of the JavaScript

Reply:
What about: 
> While synchronous JavaScript is being downloaded and executed, the HTML parser is blocked